File: practice_java_program/notpad/PossibleStringCombinations.java
public class PossibleStringCombinations{

    public static void main(String[] args) {
        String str = "ABC";
        combine(str, "", 0);
    }

    public static void combine(String str, String result, int index) {

        if (index == str.length()) {
            if (!result.isEmpty()) {
                System.out.println(result);
            }
            return;
        }

        // Include current character
        combine(str, result + str.charAt(index), index + 1);

        // Exclude current character
        combine(str, result, index + 1);
    }
}

File: python_oriented_interview/notpad/PossibleStringCombinations.py
def combine(str_val, result, index):
    if index == len(str_val):
        if result:
            print(result)
        return

    # Include current character
    combine(str_val, result + str_val[index], index + 1)
    
    # Exclude current character
    combine(str_val, result, index + 1)

if __name__ == "__main__":
    str_val = "ABC"
    combine(str_val, "", 0)
